dev/bus/export/dev/serial.hpp

```diff
@@ -0,0 +1,258 @@
+    static constexpr uint8_t xfer_scheduled_ = 0x01;
+    static constexpr uint8_t xfer_deferred_ = 0x01;
+
+    static std::atomic<uint8_t> xfer_state_;
+};
+
+template <class PBus, class SerialConfig>
+uint8_t serial<PBus, SerialConfig>::input_buffer_[SerialConfig::in_capacity]{};
+
+template <class PBus, class SerialConfig>
+uint8_t serial<PBus, SerialConfig>::output_buffer_[SerialConfig::out_capacity]{};
+
+template <class PBus, class SerialConfig>
+std::atomic_int serial<PBus, SerialConfig>::input_write_end_{};
+
+template <class PBus, class SerialConfig>
+std::atomic_int serial<PBus, SerialConfig>::input_read_end_{};
+
+template <class PBus, class SerialConfig>
+size_t serial<PBus, SerialConfig>::input_current_read_size_ = 0;
+
+template <class PBus, class SerialConfig>
+bool serial<PBus, SerialConfig>::is_inited_ = false;
+
+template <class PBus, class SerialConfig>
+std::atomic<uint8_t> serial<PBus, SerialConfig>::xfer_state_{xfer_initial_};
+
+template <class PBus, class SerialConfig>
+err serial<PBus, SerialConfig>::init()
+{
+    if (is_inited_) {
+        return err::ok;
+    }
+    auto result = PBus::init();
+    if (!is_ok(result)) {
+        return result;
+    }
+    PBus::set_handler(bus_handler);
+    result = PBus::enable_listen_mode();
+    if (!is_ok(result)) {
+        return result;
+    }
+    result = schedule_xfer();
+    if (!is_ok(result)) {
+        return result;
+    }
+    is_inited_ = true;
+    return result;
+}
+
+template <class PBus, class SerialConfig>
+void serial<PBus, SerialConfig>::bus_handler(bus_channel ch,
+        bus_event type, size_t total)
+{
+    if (ch == bus_channel::rx) {
+            if (type == bus_event::err) {
+                // TODO: smarter error handling?
+                return;
+            }
+            // It is not possible to half-transfer one byte
+            ecl_assert(type != bus_event::ht);
+            input_write_end_++;
+
+            // Transfer has been finished
+            // Schedule new xfer if possible
+            if (total == input_current_read_size_) {
+                // Wrap the write end
+                auto sz = config::in_capacity;
+                input_write_end_.compare_exchange_weak(sz, 0);
+                schedule_xfer(); // TODO: handle xfer errors
+            }
+    }
+    // TODO: handle TX events
+}
+
+template <class PBus, class SerialConfig>
+err serial<PBus, SerialConfig>::schedule_xfer()
+{
+    if (xfer_state_ == xfer_initial_) {
+        // Upon the first time read and write end
+        // are equal, and the entire buffer is available
+        input_current_read_size_ = config::in_capacity;
+    } else {
+        input_current_read_size_ = input_write_end_ > input_read_end_ ?
+            config::in_capacity - input_write_end_ :
+            input_read_end_ - input_write_end_;
+    }
+    // No space in buffer, defer xfer until next time
+    if (input_current_read_size_ == 0) {
+        xfer_state_ = xfer_deferred_;
```

Comment:
I propose not to keep old data, but rather re-write it with new one. It means that new xfer started every time old is finished.

Now, you can create a flag, meaning `reading in progress` that will signify that buffer is in use. If so, xfer must be started from main thread when it will finish its work, rather than from ISR thread.